lib/utils/packaging/unarchive_unix.go

```diff
@@ -0,0 +1,167 @@
+//go:build !windows
+
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package packaging
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"errors"
+	"io"
+	"log/slog"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"slices"
+	"strings"
+
+	"github.com/google/renameio/v2"
+	"github.com/gravitational/trace"
+)
+
+// ReplaceToolsBinaries extracts the package into `extractDir` and replaces the specified
+// applications with symlinks in the tool's directory.
+func ReplaceToolsBinaries(toolsDir string, archivePath string, extractDir string, apps []string) error {
+	switch runtime.GOOS {
+	case "darwin":
+		return replacePkg(toolsDir, archivePath, extractDir, apps)
+	default:
+		return replaceTarGz(toolsDir, archivePath, extractDir, apps)
+	}
+}
+
+func replaceTarGz(toolsDir string, archivePath string, extractDir string, apps []string) error {
+	f, err := os.Open(archivePath)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer f.Close()
+
+	gzipReader, err := gzip.NewReader(f)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	tarReader := tar.NewReader(gzipReader)
+	for {
+		header, err := tarReader.Next()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		// Skip over any files in the archive that are not in apps.
+		if !slices.ContainsFunc(apps, func(s string) bool {
+			return strings.HasSuffix(header.Name, s)
+		}) {
+			if _, err := io.Copy(io.Discard, tarReader); err != nil {
```

Comment:
Is this so that the gzipReader reads to the end and validates on Close? Have you tested that it's necessary, given the docs for Next():
> Any remaining data in the current file is automatically discarded. At the end of the archive, Next returns the error io.EOF.

Reply:
checked, works well without discard